### PR TITLE
Update dataset construction code to handle more input types.

### PR DESCRIPTION
**Pull request by MarkDaoust**
## Description of the change

Let `create_tuned_model` accept column-oriented dictionaries, file paths, and URLs.

## Motivation

Make it easy to pass more objects as datasets.

b/297597494

## Type of change

Choose one: Feature request

### CODE REVIEW

**Comment by MarkDaoust** [comment]
@xinyij-goo

File: google/generativeai/types/model_types.py
Comment (markmcd) [inline google/generativeai/types/model_types.py:212]: Any reason to use pandas vs e.g. `csv.DictReader` from stdlib?  The pandas dep is "hidden" so becomes a potential source of friction.

File: google/generativeai/types/model_types.py
Comment (markmcd) [inline google/generativeai/types/model_types.py:214]: MakerSuite supports CSVs and Sheets without header rows - can we do that here too?

Not a big deal I guess - the error handling looks pretty thorough, so it should be clear to users what the issue is. Maybe just add a note to the `create_tuned_model` docstring saying a header is required?

File: google/generativeai/types/model_types.py
Comment (markmcd) [inline google/generativeai/types/model_types.py:225]: I don't think it will work for the parsing logic here, but  `gspread` can return `sheets.googleapis.com` URLs (see cl/526209504).

No action, just an FYI in case it shows up here too.

**Review by markmcd — APPROVED** [review]
LG, and thanks!

(don't feel like you have to wait for me to re-review any minor changes either, I think GitHub will still let admins submit after the first approval)

File: google/generativeai/types/model_types.py
Comment (MarkDaoust) [reply google/generativeai/types/model_types.py:198]: https://devguide.python.org/versions/

2 years.

File: google/generativeai/types/model_types.py
Comment (MarkDaoust) [reply google/generativeai/types/model_types.py:192]: If I want to support csv and json.. how do we handle these? Try to parse it as json, and then fallback to csv?

`[1,2,3]` could parse as either.

Expect try one format (csv?), and allow a `format="json"` argument?

Comment (markmcd) [reply google/generativeai/types/model_types.py:192]: Good point.

An argument sounds good - otherwise we're still just making guesses or assumptions (e.g. the file handle is seekable, file fits in memory).

Just an idea anyway - if it's too ambiguous then let's not make it confusing.

File: google/generativeai/types/model_types.py
Comment (markmcd) [reply google/generativeai/types/model_types.py:212]: Oh so it does. I missed that. Makes sense to me.

Comment (MarkDaoust) [reply google/generativeai/types/model_types.py]: But let's stick to std-lib > fixed in [70f9070](https://github.com/google/generative-ai-python/pull/79/commits/70f9070e1e055af4d383d0cdbd13bb060ab81b4b).

File: google/generativeai/types/model_types.py
Comment (MarkDaoust) [reply google/generativeai/types/model_types.py]: > MakerSuite supports CSVs and Sheets without header rows - can we do that here too?

How do we know which column to use?

csv.DictReader requires a header.

**Review by markmcd — APPROVED** [review]

**Pull request merged**
